Implement Exclude Patterns for Snapshot- and Repository Names in Get Snapshots API

original-brownbear opened the pull request:
It's in the title. Adds support for exclude patterns combined with wildcard requests
similar to what we support for index names.

original-brownbear committed:
Implement Exclude Patterns for Snapshot- and Repository Names in Get Snapshots API

It's in the title. Adds support for exclude patterns combined with wildcard requests
similar to what we support for index names.

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

tlrx commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -60,14 +60,15 @@ Use the get snapshot API to return information about one or more snapshots, incl
 (Required, string)
 Comma-separated list of snapshot repository names used to limit the request.
-Wildcard (`*`) expressions are supported.
+Wildcard (`*`) expressions are supported including combining wildcards with exclude patterns starting in `-`.
```
```suggestion
Wildcard (`*`) expressions are supported including combining wildcards with exclude patterns starting with `-`.
```

tlrx commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -60,14 +60,15 @@ Use the get snapshot API to return information about one or more snapshots, incl
 (Required, string)
-Comma-separated list of snapshot names to retrieve. Also accepts wildcards (`*`).
+Comma-separated list of snapshot names to retrieve. Also accepts wildcards (`*`) as well as combining wildcards and exclude patterns
+starting in `-`.
```
```suggestion
starting with `-`.
```

Hopefully this is correct in English :)

original-brownbear replied:
I think both work, but I adjusted to `with` now, seems more standard :)

tlrx commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -208,6 +209,67 @@ public void testPaginationRequiresVerboseListing() throws Exception {
+            GetSnapshotsRequest.SortBy.REPOSITORY,
+            order
+        );
+        assertThat(allSorted.subList(0, allSnapshotNamesWithoutOther.size()), is(allSortedWithoutOther));
```
nit: I'd expect the other way around, checking that `allSortedWithoutOther` is equal to `allSorted.subList(0, allSnapshotNamesWithoutOther.size())` but it's not really important

original-brownbear replied:
Still looks nicer :) Adjusted it that way

tlrx commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -208,6 +209,67 @@ public void testPaginationRequiresVerboseListing() throws Exception {
+            "-" + otherPrefixSnapshot1,
+            "-" + otherPrefixSnapshot2
+        );
+        assertThat(allInOtherWithoutOtherExplicit, is(allInOther));
```
Should we test the result when a snapshot is both included and excluded?

original-brownbear replied:
++ added explicit check for this case

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/TransportGetRepositoriesAction.java:
```diff
@@ -83,18 +84,33 @@ protected void masterOperation(
+            final Set<String> repositoriesNotToGet = new HashSet<>();
             for (String repositoryOrPattern : repoNames) {
-                if (Regex.isSimpleMatchPattern(repositoryOrPattern) == false) {
+                if (seenWildcard && repositoryOrPattern.length() > 1 && repositoryOrPattern.startsWith("-")) {
```
I suppose we could iterate 2 times over the repositories, the first time to match repositories against include patterns ad the second to filter out matching repos with exclude patterns? Anyway the current code works too.

henningandersen replied:
Can we perhaps do this similar to the slm-change, where we first derive the include/exclude patterns and then run through the repos? And then extract the logic to a shared method to use such that we are sure that the include/exclude logic works the same.

tlrx reviewed (approved):
LGTM - I left minor comments, feel free to address

arteam commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/TransportGetRepositoriesAction.java:
```diff
@@ -83,18 +84,33 @@ protected void masterOperation(
+            boolean seenWildcard = false;
+            final Set<String> repositoriesToGet = new LinkedHashSet<>(); // to keep insertion order
+            final Set<String> repositoriesNotToGet = new HashSet<>();
             for (String repositoryOrPattern : repoNames) {
```
Does it make sense to rename `repoNames` to `repoPatterns`?

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:
```diff
@@ -290,11 +290,26 @@ private void loadSnapshotInfos(
         } else {
             for (String snapshotOrPattern : snapshots) {
-                if (GetSnapshotsRequest.CURRENT_SNAPSHOT.equalsIgnoreCase(snapshotOrPattern)) {
+                if (seenWildcard && snapshotOrPattern.length() > 1 && snapshotOrPattern.startsWith("-")) {
```
similarly to the repo-name matching, can we derive the include/exclude pattern lists first and share most of the logic between the two (and slm-policy)?

original-brownbear replied:
++ makes sense, the logic isn't exactly the same because of the concrete snapshot name -> not found -> throw case but I'll dry it up. I'll wait for the other PR to land to continue here then :)

original-brownbear replied:
On second thought, no we can't use the same logic as there's a difference between how the resolution works for policies and repositories/snapshots. For repositories and snapshots, if we have a concrete/non-wildcard pattern in the mix, we throw a `404`ish exception if we can't find the request repo/snapshot. We don't do any such thing for SLM policies which never cause exceptions to be thrown (which makes sense IMO). I'll try to simplify the logic a little regardless :)

henningandersen replied:
I would think we could extract the include/exclude patterns in one method shared between multiple purposes? And then run through them afterwards handling the exceptions etc.

original-brownbear replied:
The thing is, we have different behavior between the repos API (where we loop over the patterns and then nest the repos under them) while we can do the more efficient way of going over all the snapshot names and then matching against all patterns in one go (because we order the repos according to the order of the patterns instead of by their name or so).
It doesn't really seem like there's a simple way of drying this up but then again note that we're going to move the snapshot name logic away from running on heap like this in a follow-up, so I'm not sure there's too much point in abstracting things away only for that abstraction to become redundant in a week?

arteam commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/TransportGetRepositoriesAction.java:
```diff
@@ -83,18 +84,33 @@ protected void masterOperation(
             return repositories.repositories();
         } else {
-            Set<String> repositoriesToGet = new LinkedHashSet<>(); // to keep insertion order
+            boolean seenWildcard = false;
```
Do even need to have the `seenWildcard` variable? If we remove it, the tests passes. Is it used just as a perfomance optimization not to calculate negative patterns if there are no wildcards?

original-brownbear replied:
Added tests now that would catch this :) It's needed to correctly handle the case where we request stuff that is named starting in a `-`.

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -244,9 +306,17 @@ private static void assertStablePagination(String repoName, Collection<String> a
         GetSnapshotsRequest.SortBy sortBy,
-        SortOrder order
+        SortOrder order,
+        String... extraPatterns
```
I think this is the name patterns, could we call it `namePatterns`?

original-brownbear replied:
Renamed :)

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -208,6 +209,67 @@ public void testPaginationRequiresVerboseListing() throws Exception {
+            patternOtherRepo,
+            GetSnapshotsRequest.SortBy.REPOSITORY,
+            order,
+            "-" + otherPrefixSnapshot1,
```
nit: it seems odd that for the repo pattern we pass in a comma-separated string, whereas for the names we pass in a list of name patterns. I would prefer to see this harmonized.

original-brownbear replied:
harmonized to using arrays/var-arg for both :)

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -208,6 +209,67 @@ public void testPaginationRequiresVerboseListing() throws Exception {
         );
     }
 
+    public void testExcludePatterns() throws Exception {
```
Can we add a test validating that using `-xyz` finds such a snapshot (and repo) if no wildcard is specified first, i.e., at that we can still find snapshots/repos starting with `-`?

original-brownbear replied:
Good point! -> added a couple of tests around this scenario now.

henningandersen reviewed (commented):
Left a few comments to consider.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
make logic for resolving things nicer

original-brownbear committed:
align pattern usage

original-brownbear committed:
test with dash

original-brownbear committed:
CR comments

original-brownbear commented:
Thanks Henning, all points addressed now (though I couldn't dry things up as requested I'm afraid).

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-tests-windows-sample

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-tests-windows-sample

original-brownbear commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into support-negative-snapshot-patterns

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/get/TransportGetRepositoriesAction.java:
```diff
@@ -80,30 +88,37 @@ protected void masterOperation(
-                if (Regex.isSimpleMatchPattern(repositoryOrPattern) == false) {
-                    repositoriesToGet.add(repositoryOrPattern);
+        }
+        final List<String> includePatterns = new ArrayList<>();
```
I think the new exclude functionality also affects the list repositories API, we should also update the documentation for that API then.

original-brownbear replied:
++ udpated

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java:
```diff
@@ -210,12 +216,151 @@ public void testPaginationRequiresVerboseListing() throws Exception {
+                )
+            )
+        );
+    }
```
Can we also verify that `non-existing*,-weird-snapshot-1` does not return `-weird-snapshot-1`? And that '*,--weird-snapshot-1` also does not return it? And that `-*` returns all `-` prefixed snapshots.

original-brownbear replied:
++ added those cases :)

henningandersen reviewed (commented):
A few smaller comments. Still hoping for a bit of code/sharing on include/exclude patterns :slightly_smiling_face:

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
tests

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
Merge remote-tracking branch 'origin/support-negative-snapshot-patterns' into support-negative-snapshot-patterns

original-brownbear commented on docs/reference/snapshot-restore/apis/get-snapshot-api.asciidoc:
```diff
@@ -60,14 +60,15 @@ Use the get snapshot API to return information about one or more snapshots, incl
 `<snapshot>`::
 (Required, string)
-Comma-separated list of snapshot names to retrieve. Also accepts wildcards (`*`).
+Comma-separated list of snapshot names to retrieve. Also accepts wildcards (`*`) as well as combining wildcards and exclude patterns
```
Right, that was pointless, made it the same for both as suggested now.

original-brownbear commented:
Thanks @henningandersen! addresed all points but sharing the pattern resolution code, to me at least it seems with the way we do the ordering and not-ordering when resolving things, it's going to be pain to share code at this point in time I think, but lets see what you think :)

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
desired concrete name + wildcard behavior

original-brownbear committed:
_current is a wildcard pattern now

original-brownbear commented:
Alright @henningandersen this should be good for another round. Implemented `_current` as wildcard as well as fixed the `pattern,name,excludePatternForName` case we discussed last night and added tests for both :)

henningandersen reviewed (approved):
LGTM.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into support-negative-snapshot-patterns

original-brownbear committed:
missing sentence

original-brownbear commented:
Thanks everyone!
